src/internal/clusterstate/v2.5.0/util.go

```diff
@@ -0,0 +1,57 @@
+package v2_5_0
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	"go.uber.org/zap"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+)
+
+type postgresBatcher struct {
+	ctx   context.Context
```

Comment:
Generally one shouldn’t stash a context in a structure; accept it as a parameter instead. The reason for this is that the context is meant to be carried along the dynamic contour. So:

```go
func newPostgresBatcher(ctx context.Context, tx *pachsql.Tx) *postgresBatcher { … }

func (pb *postgresBatcher) Add(ctx context.Context, stmt string) error {
	⋮
	return pb.execute(ctx)
}

func (pb *postgresBatcher) execute(ctx context.Context) error { … }

func (pb *postgresBatcher) Close(ctx context.Context) error {
	return pb.execute(ctx)
}
```

Passing a context to a close method may be controversial, but I think in this case it’s reasonable.